### PR TITLE
Fix wrong Org for ProductPlanning when created from ProductPlanningSchema

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/9958

### CODE REVIEW

**Review by metas-ts — APPROVED** [review]

File: backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java [hunk 1]
```diff
@@ -335,9 +342,9 @@ private void createOrUpdateProductPlanningsForSelector()
 		}
 		
 		final ProductId productId = ProductId.ofRepoId(getM_Product_ID());
+		final OrgId orgId = OrgId.ofRepoId(getAD_Org_ID());
 
-		final IProductPlanningSchemaBL productPlanningSchemaBL = Services.get(IProductPlanningSchemaBL.class);
-		productPlanningSchemaBL.createOrUpdateProductPlanningsForSelector(productId, productPlanningSchemaSelector);
+		productPlanningSchemaBL.createOrUpdateProductPlanningsForSelector(productId, orgId, productPlanningSchemaSelector);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java:347]: i would suggest to extract that code in an model interceptor.

As a side benefit we will not wire"private final IProductPlanningSchemaBL productPlanningSchemaBL"  service each time a product is loaded.

File: backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaBL.java [hunk 2]
```diff
@@ -61,28 +65,30 @@
...
 
-		return createOrUpdateProductPlanningsForSelector(productId, selector);
+		return createOrUpdateProductPlanningsForSelector(productId, orgId, selector);
 	}
 
 	@Override
 	public List<I_PP_Product_Planning> createOrUpdateProductPlanningsForSelector(
-			@NonNull final ProductId productId,
-			@NonNull final ProductPlanningSchemaSelector selector)
+			final @NonNull ProductId productId,
```
[the remark this answers is not in the document]
Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaBL.java:83]: unfortunately not available, but there is a feature request (with duplicates): https://youtrack.jetbrains.com/issue/IDEA-168035

File: backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java (hunk 1 above)
Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java:347]: This code is called only once after the product was created, afaiu from 

```java
protected boolean afterSave(boolean newRecord, boolean success) {
[...]
	if(newRecord)
		{
			createOrUpdateProductPlanningsForSelector();
		}
```
so isnt' this already in Product's model interceptor?

File: backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java [hunk 3]
```diff
@@ -70,23 +80,82 @@ public static ProductPlanningSchema getById(@NonNull final ProductPlanningSchema
...
+	 * Used as a delegate to ProductPlanningSchema, to calculate if 2 schemas are equals, ignoring only their ID and Org
+	 */
+
+	@Value
+	private static class ProductPlanningSchemaIgnoringOrgAndId
+	{
+		ProductPlanningSchema schema;
+
+		@Override
+		public boolean equals(final Object o)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java:129]: this is too fragile IMHO

Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java:129]: what do you mean? Why is it fragile?

File: backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java [hunk 4]
```diff
@@ -154,7 +223,7 @@ public static void save(@NonNull final ProductPlanningSchema schema)
 	/**
 	 * @return All the active products with the given product planning schema selector
 	 */
-	public static Set<ProductId> retrieveProductIdsForSchemaSelector(
+	public static Set<ImmutablePair<ProductId, OrgId>> retrieveProductIdsForSchemaSelector(
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java:226]: consider moving this method to ProductDAO

Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java:226]: done

File: backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java (hunk 1 above)
Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java-legacy/org/compiere/model/MProduct.java:347]: done

File: backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java (hunk 3 above)
Comment (TheBestPessimist) [reply backend/de.metas.business/src/main/java/de/metas/product/impl/ProductPlanningSchemaDAO.java]: updated after functional change.

**Pull request merged**